src/io.rs

```diff
@@ -57,84 +85,356 @@ fn spawn_input_stream(
+                    return match l {
+                        LatencyHint::Interactive => default_buffer_size,
+                        LatencyHint::Balanced => match self.supported.buffer_size() {
+                            SupportedBufferSize::Range { max, .. } => {
+                                let b = (default_buffer_size * 2).min(*max);
+                                (b + BUFFER_SIZE - 1) / BUFFER_SIZE * BUFFER_SIZE
+                            }
+                            SupportedBufferSize::Unknown => {
+                                let b = default_buffer_size * 2;
+                                (b + BUFFER_SIZE - 1) / BUFFER_SIZE * BUFFER_SIZE
+                            }
+                        },
+                        LatencyHint::Playback => match self.supported.buffer_size() {
+                            SupportedBufferSize::Range { max, .. } => {
+                                let b = (default_buffer_size * 4).min(*max);
+                                (b + BUFFER_SIZE - 1) / BUFFER_SIZE * BUFFER_SIZE
+                            }
+                            SupportedBufferSize::Unknown => {
+                                let b = default_buffer_size * 4;
+                                (b + BUFFER_SIZE - 1) / BUFFER_SIZE * BUFFER_SIZE
+                            }
+                        },
+                        // b is always positive
+                        #[allow(clippy::cast_sign_loss)]
+                        // truncation is the desired behavior
+                        #[allow(clippy::cast_possible_truncation)]
+                        LatencyHint::Specific(t) => {
+                            let b = t * f64::from(self.prefered.sample_rate.0);
+                            (b as u32 + BUFFER_SIZE - 1) / BUFFER_SIZE * BUFFER_SIZE
+                        }
+                    };
+                }
+            },
+            None => (default_buffer_size + BUFFER_SIZE - 1) / BUFFER_SIZE * BUFFER_SIZE,
+        }
+    }
 
-    // communication channel to the render thread
-    let (mut sender, receiver) = crossbeam_channel::unbounded();
+    /// modify the sample rate config, following user options
+    ///
+    /// # Argument
+    ///
+    /// * `options` - options contains sample rate information
+    fn with_sample_rate(mut self, options: Option<&AudioContextOptions>) -> Self {
+        if let Some(opts) = options {
+            if let Some(fs) = opts.sample_rate {
+                self.prefered.sample_rate.0 = fs;
+            }
+        };
+
+        self
+    }
 
-    // spawn the render thread
-    let frames_played_clone = frames_played.clone();
-    let render = RenderThread::new(
-        sample_rate,
-        channels as usize,
-        receiver,
-        frames_played_clone,
-    );
+    /// modify the buffer size config, following user options
+    ///
+    /// # Argument
+    ///
+    /// * `options` - options contains latency hint information
+    ///
+    /// # Warning
```

Comment:
As mentioned here the correctness of the API is not enforce by the type system. I could make it more strict but at the cost of more complexity. What do you think ?